geode-core/src/main/java/org/apache/geode/internal/cache/BucketRedundancyTracker.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.cache;
+
+/**
+ * Keeps track of redundancy status for a bucket in a PartitionedRegion and update the region's
+ * {@link PartitionedRegionRedundancyTracker} of the bucket's status for the region.
+ */
+class BucketRedundancyTracker {
+  private boolean redundancySatisfied = false;
+  private boolean hasAnyCopies = false;
+  private boolean redundancyEverSatisfied = false;
+  private boolean hasEverHadCopies = false;
+  private int currentRedundancy = -1;
+  private final int targetRedundancy;
+  private final PartitionedRegionRedundancyTracker regionRedundancyTracker;
+
+  /**
+   * Creates a new BucketRedundancyTracker
+   *
+   * @param redundantCopies the number of redundant copies specified for the
+   *        {@link PartitionedRegion} of this bucket
+   * @param regionRedundancyTracker the redundancy tracker for the {@link PartitionedRegion} of this
+   *        bucket
+   */
+  BucketRedundancyTracker(int redundantCopies,
+      PartitionedRegionRedundancyTracker regionRedundancyTracker) {
+    this.targetRedundancy = redundantCopies;
+    this.regionRedundancyTracker = regionRedundancyTracker;
+  }
+
+  /**
+   * Adjust statistics based on closing a bucket
+   */
+  synchronized void closeBucket() {
+    if (!redundancySatisfied) {
```

Comment:
Since "redundancySatisified" starts out false what prevents closeBucket from decrementing the stat when it has never incremented it? Same question for "hasAnyCopies". I can't tell that we will always increment these stats before closeBucket is called.